peer/internal/uniform/rand.go

```diff
@@ -0,0 +1,202 @@
+// Copyright (c) 2024 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+//
+// Uniform random algorithms modified from the Go math/rand/v2 package with
+// the following license:
+//
+// Copyright (c) 2009 The Go Authors. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+//    * Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//    * Redistributions in binary form must reproduce the above
+// copyright notice, this list of conditions and the following disclaimer
+// in the documentation and/or other materials provided with the
+// distribution.
+//    * Neither the name of Google Inc. nor the names of its
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+// Package uniform provides uniformly distributed, cryptographically secure
+// random numbers with randomness obtained from a crypto/rand.Reader or other
+// CSPRNG reader.
+//
+// Random sources are required to never error; any errors reading the random
+// source will result in a panic.
+package uniform
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+	"math/bits"
+	"time"
+)
+
+func read(rand io.Reader, buf []byte) {
+	_, err := io.ReadFull(rand, buf)
+	if err != nil {
+		panic(fmt.Errorf("uniform: read of random source errored: %w", err))
```

Comment:
I know that this pretty much can't fail on modern OSes, but I really don't want panics on this in peer.  dcrd being taken down in the event of an, albeit unlikely, but otherwise routine error is unacceptable.  Perhaps returning a success bool and falling back to something else is a solution that doesn't require passing an error throughout everything, but the panic is a no go.

Reply:
I thought about possible fallbacks we could use and failed to come up with any while still keeping the cryptographically-secure guarantees.  We could punt to using math/rand, but I would very much rather not have to do that.  I have plans to use this same code elsewhere where the CS guarantees are absolutely required and didn't want to need to maintain multiple different variants of this.  (math/rand is also very slow, which is another reason it should never be used.)

This whole package ideally wouldn't exist and we would use math/rand/v2 with our own Source implementation.  The Source only returns uniformly random uint64s, with no ability to return an error.  They don't outright say it, but it sure does look like that interface is designed with crypto/rand-based implementations panicking or otherwise killing the process during any exceptional error.  Alas, math/rand/v2 is new to Go 1.22 and we can't use it here.

With that out of the way let's consider what OSes and versions might still error from a crypto/rand read:

* On Windows, crypto/rand will always read [ProcessPrng](https://learn.microsoft.com/en-us/windows/win32/seccng/processprng).  This is a per-CPU userspace CSPRNG that Windows provides to all Win32 processes.  It never errors.  (Also, it would be interesting to see benchmarks, but I bet we can just use crypto/rand directly on Windows without overdoing it with our own uprng package).
* On macOS, OpenBSD, and NetBSD, crypto/rand always reads randomness with [getentropy](https://man.openbsd.org/getentropy).  It never errors.
* On Linux, FreeBSD, DragonFly BSD, and Solaris, randomness is obtained from [getrandom](https://www.man7.org/linux/man-pages/man2/getrandom.2.html).  It never errors when only filling a buffer of up to 32-bytes, as we do here.  So, let's consider the versions that might not have getrandom:
  * On Linux, getrandom was introduced by kernel 3.17.  The kernel.org project does not maintain any longterm kernel branches of this or earlier versions, so let's consider vendor kernels.  [All of Ubuntu's still-maintained LTS branches are rebased on Linux 4.4 or newer](https://ubuntu.com/about/release-cycle#ubuntu-kernel-release-cycle).  Debian 10 "buster" is the oldest still-maintained LTS release and it [rebased on Linux 5.10](https://packages.debian.org/buster/kernel/linux-image-5.10-amd64).  I can't find anything officially from Red Hat anymore but Centos 7 (corresponding to RHEL 7, the oldest still-maintained RHEL branch) was [initially released with Linux 3.10](https://tech.joellemena.com/centos/centos-7-kernel-version-list/).
  * On FreeBSD, [getrandom was introduced in 12.0-RELEASE](https://man.freebsd.org/cgi/man.cgi?query=getrandom&sektion=2&n=1), which was released in December 2018.  FreeBSD 12-STABLE branch is [no longer officially supported](https://www.freebsd.org/releng/).
  * On DragonFly, [getrandom was introduced in 5.8](https://www.dragonflybsd.org/release58/), released in March 2020.  The 5.8 branch [doesn't appear to be maintained anymore](https://gitweb.dragonflybsd.org/dragonfly.git/shortlog/refs/heads/DragonFly_RELEASE_5_8).
  * On Illumos, [getrandom was committed](https://github.com/illumos/illumos-gate/commit/9d12795f87b63c2e39e87bff369182edd34677d3) in April 2015.  I'm not aware of any current Illumos distribution that is still rebased over even older versions of gate.
  * On Oracle Solaris, [getrandom was introduced in 11.3](https://blogs.oracle.com/solaris/post/solaris-new-system-calls-getentropy2-and-getrandom2), released October 2015.  [Previous releases of Solaris 11 are no longer maintained](https://endoflife.date/oracle-solaris).

I'm not concerned any of the above operating systems ever erroring during a crypto/rand read.

There is one unfortunate exception I found, and that is iOS.  Apparently, `getentropy` is a private interface on non-macOS Darwin, so it will fall back to reading /dev/urandom.  However, the common cases that one might expect a read of /dev/urandom to fail (a closed file descriptor, or the file missing from a chroot) are even less likely to occur on iOS.  Go should probably handle this specially by using 

A less unfortunate exception is Plan9.  It reads /dev/random and there is the possibility that the read fails, but I am not clear on all the ins and outs of what exactly may cause errors on that platform.  I say this is less unfortunate because we don't officially support running on this platform, although dcrd does compile for it (dcrwallet, however, does not).

The only other Unix OS I can find that doesn't get special syscall treatment is AIX, but dcrd doesn't compile there:

```
$ GOOS=aix GOARCH=ppc64 go build
# github.com/syndtr/goleveldb/leveldb/storage
../../go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20210819022825-2ae1ddf74ef7/leveldb/storage/file_storage.go:107:16: undefined: newFileLock
../../go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20210819022825-2ae1ddf74ef7/leveldb/storage/file_storage.go:192:3: undefined: rename
../../go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20210819022825-2ae1ddf74ef7/leveldb/storage/file_storage.go:267:12: undefined: rename
../../go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20210819022825-2ae1ddf74ef7/leveldb/storage/file_storage.go:272:12: undefined: syncDir
../../go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20210819022825-2ae1ddf74ef7/leveldb/storage/file_storage.go:555:9: undefined: rename
../../go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20210819022825-2ae1ddf74ef7/leveldb/storage/file_storage.go:591:13: undefined: syncDir
```

---

So, given all that information, if a crypto/rand read did somehow still error, I would think the responsible thing to do would be to panic or otherwise kill the process (e.g. write error message to stderr and call os.Exit).  It would not be a "routine" error that we are dealing with anymore; there is some serious bug affecting either the process itself, or the operating system, and possibly indicates an attacker already has or is trying to take root.  I can't in good conscious allow the process to continue executing as if nothing is wrong.

Reply:
Ok but, what to use then instead? I'm not sure what "using the stdlib for the delay selection logic" means, when these packages *are* just using the stdlib for exactly that purpose.